pkg/agent/nodeportlocal/k8s/npl_controller.go

```diff
@@ -210,11 +210,19 @@ func (c *NPLController) enqueueSvcUpdate(oldObj, newObj interface{}) {
 		} else if newSvcAnnotation == "true" {
 			podKeys = sets.NewString(c.getPodsFromService(newSvc)...)
 		}
-	} else if !reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector) {
+	}
+
+	var oldPodSet, newPodSet sets.String
+	if oldSvc.Spec.Type != newSvc.Spec.Type {
+		newPodSet = sets.NewString(c.getPodsFromService(newSvc)...)
+		podKeys = podKeys.Union(newPodSet)
+	}
+
+	if !reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector) {
 		// Disjunctive union of Pods from both Service sets.
-		oldPodSet := sets.NewString(c.getPodsFromService(oldSvc)...)
-		newPodSet := sets.NewString(c.getPodsFromService(newSvc)...)
-		podKeys = oldPodSet.Difference(newPodSet).Union(newPodSet.Difference(oldPodSet))
+		oldPodSet = sets.NewString(c.getPodsFromService(oldSvc)...)
+		newPodSet = sets.NewString(c.getPodsFromService(newSvc)...)
+		podKeys = podKeys.Union(oldPodSet.Difference(newPodSet).Union(newPodSet.Difference(oldPodSet)))
```

Comment:
could we write our own `SymmetricDifference` method instead of doing `oldPodSet.Difference(newPodSet).Union(newPodSet.Difference(oldPodSet))`? It will be more readable and it may also be more efficient since we would not be building intermediate sets.

Reply:
+1, maybe the difference is minor here, but I found using Union when calculating the union of many sets affected the performance a lot (in particular, the TestInitXLargeScaleWithOneNamespace test) as it always makes a new set and copies all items when merging each set. I added a function to improve it in https://github.com/vmware-tanzu/antrea/pull/1938. Maybe we could have a separate PR to add `SymmetricDifference` in same place and replace all such usage.